bin.src/genDefectList.py

```diff
@@ -0,0 +1,265 @@
+    logger.debug('uniqueFirstIntsSorted = %s' % uniqueFirstIntsSorted)
+    for a in uniqueFirstIntsSorted:
+        if a < ccd:
+            for i in range(len(textArr)):
+                if firstInts[i] == a:
+                    text_file.write(textArr[i])
+
+allFlatsLow = np.ndarray(shape=[image.shape[0], image.shape[1], len(flatsLow)], dtype='float32')
+allFlatsHigh = np.ndarray(shape=[image.shape[0], image.shape[1], len(flatsHigh)], dtype='float32')
+for i in np.arange(len(flatsLow)):
+    dataId = dict(visit=flatsLow[i], spectrograph=args.spectrograph, arm=args.arm)
+    image = butler.get('postISRCCD', dataId).getMaskedImage().getImage().getArray()
+    allFlatsLow[:,:,i] = image
+for i in np.arange(len(flatsHigh)):
+    dataId = dict(visit=flatsHigh[i], spectrograph=args.spectrograph, arm=args.arm)
+    image = butler.get('postISRCCD', dataId).getMaskedImage().getImage().getArray()
+    allFlatsHigh[:,:,i] = image
+
+nLow = len(flatsLow)
+if nLow > 2:
+    medianFlatsLow = np.median(allFlatsLow,2)
+    logger.info('%d Flats with %ds exposure time found' % (nLow, args.expTimeLow))
+elif nLow == 2:
+    medianFlatsLow = np.min(allFlatsLow,2)
+    logger.warn('Only %d Flats with %ds exposure time found' % (nLow, args.expTimeLow))
+elif nLow < 1:
+    raise RunTimeError("No Flats with %d s exposure time found" % args.expTimeLow)
+else:
+    logger.warn('Only 1 Flat with %ds exposure time' % args.expTimeLow)
+    medianFlatsLow = allFlatsLow[:,:,0]
+
+nHigh = len(flatsHigh)
+if nHigh > 2:
+    medianFlatsHigh = np.median(allFlatsHigh,2)
+    logger.info('%d Flats with %ds exposure time found' % (nLow, args.expTimeHigh))
+elif nHigh == 2:
+    medianFlatsHigh = np.min(allFlatsHigh,2)
+    logger.warn('Only %d Flats with %ds exposure time found' % (nLow, args.expTimeHigh))
+elif nHigh < 1:
+    raise RunTimeError("No flats with %d s exposure time found" % args.expTimeHigh)
+else:
+    logger.warn('Only 1 Flat with %ds exposure time' % args.expTimeHigh)
+    medianFlatsHigh = allFlatsHigh[:,:,0]
+
+medianFlatsLow = drpStella.where( medianFlatsLow, '==', 0., 0.000001, medianFlatsLow)
+medianFlatsHigh = drpStella.where( medianFlatsHigh, '==', 0., 0.000001, medianFlatsHigh)
+divFlat = medianFlatsHigh / medianFlatsLow
+
+sctrl = afwMath.StatisticsControl()
+sctrl.setWeighted(False)
+
+divFlatIm = afwImage.ImageF(divFlat)
+divFlatMIm = afwImage.makeMaskedImage(divFlatIm)
+
+mean = afwMath.makeStatistics(divFlatMIm, afwMath.MEANCLIP).getValue()
+stddev = afwMath.makeStatistics(divFlatMIm,afwMath.STDEVCLIP).getValue()
+
+if args.outFile != '':
+    mask = divFlatMIm.getMask()
+    maskVal = 1 << divFlatMIm.getMask().getMaskPlane('BAD')
+    for badCol in badCols:
+        text_file.write("%d %d 0 1 %d\n" % (ccd, badCol, args.nRows))
+        mask[badCol,:] |= maskVal
+    nBad = 0
+    for iRow in np.arange(args.nPixCut,divFlat.shape[0]-args.nPixCut):
+        for iCol in np.arange(args.nPixCut,args.gapLow):
+            if iCol not in badCols:
```

Comment:
badCols are written to the defect list before checking for individual bad pixels. As the bad columns are already in the defect list we only check pixels not in these bad columns.
I'm not sure why these columns are bad. It doesn't seem to be incorrect geometry. I did ask Craig but he said he never cared about these columns as they won't contain data anyway.